### DISCUSSION TITLE
关于语法树输出是并列的兄弟节点

### DISCUSSION THREAD

**Discussion by Fiona-ZD** [arcsysu/SYsU-lang#27] [Q&A]
第二个实验中，从%start CompUnit开始解析。请问对于既有全局变量，又有main函数的情况，这两个在输出时应该是并列的关系，但是好像只能有一个start入口，这种要怎么处理呢？也就是在补充的例子当中，一个节点下面有两个子节点的情况。

**Comment by wu-kan** [comment]
可以参照 `Block`，我写了多个匹配规则，即可以匹配空块，也可以匹配有内容的块。你应该也可以通过给 `CompUnit` 写多个匹配来实现目标？


```yacc
Block: T_L_BRACE T_R_BRACE {}

Block: T_L_BRACE BlockItem T_R_BRACE {}
```

**Reply by Fiona-ZD** [reply]
但是start好像只能匹配一个规则？我给%start CompUnit写多个匹配规则，但是好像只能有一个作为入口。在既有全局变量，又有main函数的输出当中，和CompUnit: ValDeclare匹配上之后，就不会再和CompUnit: FunctionDecl匹配了。换句话说，我想问的是，怎么像下面这样，VarDecl和FunctionDecl作为同一级输出？
"inner": [
    {
      "id": "0x1282d30",
      "kind": "VarDecl",
      "isUsed": true,
      "name": "a",
      "mangledName": "a",
      "type": {
        "qualType": "int"
      }
    },
    {
      "id": "0x1282e38",
      "kind": "FunctionDecl",
      "name": "main",
      "mangledName": "main",
      "type": {
        "qualType": "int ()"
      },
    }
  ]

**Reply by wu-kan** [reply]
```yacc
CompUnit: CompUnit VarDecl {}
 
CompUnit: CompUnit FuncDecl {}

CompUnit: VarDecl {}

CompUnit: FuncDecl {}
```

这样可以吗？

**Reply by Fiona-ZD** [reply]
我试了一下，好像不行？报了个syntax error

**Reply by wu-kan** [reply]
```yacc
%start StartUnit
%%
StartUnit: CompUnit YYEOF {}
```

再用这个替换掉原来的起始单元呢？

**Reply by Fiona-ZD** [reply]
好像还是不行？一样是syntax error。对于下面的代码，好像是匹配int a，之后的main函数没匹配。
`
int a;
int main(){
    return 3;
}
`
输出：
`
syntax error{
  "kind": "ValDecl",
  "name": "a"
}
`

**Reply by wu-kan** [reply]
参考下[科大的实验](https://github.com/Komorebi660/SysYF-Compiler/blob/master/grammar/SysYFParser.yy)是怎么写的？
